### ISSUE TITLE
Investigate performance regressions when using the changes coming from #149

### ISSUE THREAD

**Issue by mseeger** [awslabs/keys_values#152]
**Describe the bug**  
After merging the PR #149, I noted a performance regression. When running gradient computation, the GPU devices were used only up to 40% or so. Before the change, usage is 100% or close to.

The PR was reverted for now. Can use commit c3b1b22fb366a0843828b56f4fe18c06a110e6e1 to get what I was running (just before the reverts).

**To reproduce** 

I was running something closely related to this:

```
export DATASET_KEY="nq"; \
export DATASET_SIZE="64k"; \
export EVAL_STEPS=10; \
export MAX_EPOCHS=5; \
export KV_LOGIC="h2o"; \
export KVL_SHORT="h2o"; \
export KV_NORM="False"; \
export CHUNK_SIZE=128; \
export CHUNK_MULTIPLIER=0.1; \
CUDA_VISIBLE_DEVICES="4,5,6,7" \
PYTORCH_ALLOC_CONF=expandable_segments:True \
KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/rerun/helmet_${DATASET_KEY}_${DATASET_SIZE}/${KVL_SHORT}_4gpu_cs${CHUNK_SIZE}_lr5/logs" \
python3 keys_values/__main__.py finetune_long_lora \
    Qwen/Qwen3-4B-Instruct-2507 \
    --out_dir /home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/rerun/helmet_${DATASET_KEY}_${DATASET_SIZE}/${KVL_SHORT}_4gpu_cs${CHUNK_SIZE}_lr5 \
    --precision bf16-true \
    --verbose some \
    --devices 4 \
    --data Helmet \
        --data.dataset_key ${DATASET_KEY} \
        --data.max_length ${DATASET_SIZE} \
        --data.metadata_dir /home/ubuntu/out/finetune/data \
        --data.trainloader_longest_first True \
    --train.save_interval ${EVAL_STEPS} \
        --train.micro_batch_size 2 \
        --train.epochs ${MAX_EPOCHS} \
        --train.average_loss_per_batch True \
    --eval.interval ${EVAL_STEPS} \
        --eval.initial_validation False \
        --eval.use_sample_metric False \
    --kv_cache.cache_length 32768 \
        --kv_cache.chunk_size ${CHUNK_SIZE} \
        --kv_cache.name ${KV_LOGIC}-torch-quantized8 \
        --kv_cache.normalize_scores ${KV_NORM} \
    --grad.layers_per_cell 1 \
        --grad.layercp_qname default \
        --grad.cachecp_qname torch-quantized8 \
        --grad.chunks_per_cell_multiplier ${CHUNK_MULTIPLIER} \
    --optimizer.name AdamW \
        --optimizer.learning_rate 0.0005
```

I ran this on a P4 instance, using 4 of the 8 devices. I then ran `nvidia-smi`, which showed low GPU utilization (20-30%, max 70%) during gradient computation.

**Expected behavior**  
Before the PR was merged (and after the revert), GPU utilization during gradient computation is 100% or close to.

**Additional context**  
One guess: Maybe the transfer to CPU for parking buffers is blocking GPU computation. Note that for the case above, the old code works, so the new capabilities of #149 are not needed. One would have to write it such that transfer to CPU only happens if the old setup does not work, so that gap is larger than 1 step.
